homeassistant/components/http/auth.py

```diff
@@ -40,10 +43,104 @@ def async_sign_path(hass, refresh_token_id, path, expiration):
 
 
 @callback
-def setup_auth(app, trusted_networks, api_password):
+def setup_auth(hass, app):
     """Create auth middleware for the app."""
     old_auth_warning = set()
 
+    support_legacy = hass.auth.support_legacy
+    if support_legacy:
+        _LOGGER.warning("legacy_api_password support has been enabled.")
+
+    trusted_networks = []
+    for prv in hass.auth.auth_providers:
+        if prv.type == 'trusted_networks':
+            trusted_networks += prv.trusted_networks
+
+    async def async_validate_auth_header(request):
+        """
+        Test authorization header against access token.
+
+        Basic auth_type is legacy code, should be removed with api_password.
+        """
+        if hdrs.AUTHORIZATION not in request.headers:
+            return False
+
+        try:
+            auth_type, auth_val = \
+                request.headers.get(hdrs.AUTHORIZATION).split(' ', 1)
+        except ValueError:
+            # If no space in authorization header
+            return False
+
+        if auth_type == 'Bearer':
+            refresh_token = await hass.auth.async_validate_access_token(
+                auth_val)
+            if refresh_token is None:
+                return False
+
+            request[KEY_HASS_USER] = refresh_token.user
+            return True
+
+        if auth_type == 'Basic' and support_legacy:
+            decoded = base64.b64decode(auth_val).decode('utf-8')
+            try:
+                username, password = decoded.split(':', 1)
+            except ValueError:
+                # If no ':' in decoded
+                return False
+
+            if username != 'homeassistant':
+                return False
+
+            user = await legacy_api_password.async_validate_password(
+                hass, password)
+            if user is None:
+                return False
+
+            request[KEY_HASS_USER] = user
+            _LOGGER.info(
+                'Basic auth with api_password is going to deprecate,'
+                ' please use a bearer token to access %s from %s',
+                request.path, request[KEY_REAL_IP])
+            old_auth_warning.add(request.path)
+            return True
+
+        return False
+
+    async def async_validate_signed_request(request):
```

Comment:
What's the benefit of pulling these into the `setup_auth` method ? I would prefer them stay outside